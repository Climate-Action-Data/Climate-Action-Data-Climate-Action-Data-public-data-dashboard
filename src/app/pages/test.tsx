'use client'
import { type NextPage } from "next";

const Test: NextPage = () => {
    return (
        <div>
            <h1>Hello</h1>
            <p>data</p>
        </div>
    )
}

export default Test