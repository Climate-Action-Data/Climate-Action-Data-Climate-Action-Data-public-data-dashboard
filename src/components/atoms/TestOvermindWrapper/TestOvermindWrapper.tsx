import { createOvermind } from 'overmind'
import { Provider } from 'overmind-react'
import { config } from '@/overmind'
import { PropsWithChildren } from 'react'

const overmind = createOvermind(config)

export const TestOvermindWrapper = (props: PropsWithChildren) => {
  return <Provider value={overmind}>{props.children}</Provider>
}
