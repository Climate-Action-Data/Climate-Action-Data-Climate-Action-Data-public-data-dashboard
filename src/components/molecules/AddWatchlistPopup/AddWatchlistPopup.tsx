import { Flex, Button, Modal, ModalBody, ModalContent, ModalFooter, ModalOverlay, Input, InputGroup, InputLeftElement, VStack, Divider } from '@chakra-ui/react'
import { useEffect, useState } from 'react'
import { CloseIcon } from '@/components/atoms/CloseIcon/CloseIcon'
import { useActions } from '@/overmind'
import { Watchlist } from '@/@types/Watchlist'
import { useTranslation } from 'react-i18next'
import { SearchIcon } from '@/components/atoms/SearchIcon/SearchIcon'
import { PlusIcon } from '@/components/atoms/PlusIcon/PlusIcon'
import { EditWatchlistPopup } from '@/components/atoms/EditWatchlistPopup/EditWatchlistPopup'
import { WatchlistCheckbox } from '@/components/atoms/WatchlistCheckbox/WatchlistCheckbox'

interface AddWatchlistPopupProps {
  warehouseProjectId: string
  onModalClose?: () => void
  isOpen: boolean
}

export const AddWatchlistPopup = (props: AddWatchlistPopupProps) => {
  const { getAllWatchlist, createOneWatchlist, addProjectToWatchlist } = useActions().watchlist
  const { t } = useTranslation(`watchlist`)
  const { onModalClose, isOpen, warehouseProjectId } = props
  const [watchlists, setWatchlists] = useState<Watchlist[] | undefined>(undefined)
  const [visibleWatchlists, setVisibleWatchlists] = useState<Watchlist[] | undefined>(undefined)
  const [isModalOpen, setIsModalOpen] = useState<boolean>(false)
  const [showCreatePopup, setShowCreatePopup] = useState<boolean>(false)
  const [searchWatchlist, setSearchWatchlist] = useState<string>(``)

  useEffect(() => {
    getAllWatchlist()
      .then((result) => {
        if (result.data) {
          result.data.sort((a, b) => {
            return new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime()
          })
          setWatchlists(result.data)
          setVisibleWatchlists(result.data)
        }
      })
      .catch(() => undefined)

    if (!isModalOpen) {
      setIsModalOpen(isOpen)
    }
  }, [isOpen])

  const handleClose = () => {
    setIsModalOpen(false)
    onModalClose?.()
  }

  const handleNewWatchlist = () => {
    setShowCreatePopup(true)
    setIsModalOpen(false)
  }

  const handleCreateConfirm = async (name: string, description: string) => {
    const watchlist = await createOneWatchlist({ name, description })
    if (watchlist?.data?.id) {
      try {
        await addProjectToWatchlist({ warehouseProjectId, watchlistId: watchlist.data.id })
      } catch (error) {}
    }
    setShowCreatePopup(false)
    onModalClose?.()
  }

  const handleCreateCancel = () => {
    setIsModalOpen(true)
    setShowCreatePopup(false)
  }

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value
    if (value === ``) {
      setVisibleWatchlists(watchlists)
    } else {
      setVisibleWatchlists(
        watchlists?.filter((watchlist) => {
          return watchlist.name.toLowerCase().includes(value.toLowerCase())
        }),
      )
    }
    setSearchWatchlist(value)
  }

  return (
    <>
      <Modal onClose={handleClose} size={`md`} isOpen={isModalOpen}>
        <ModalOverlay />
        <ModalContent justifyContent="space-between" minH="500px">
          <Flex padding="16px" flexDirection="column" boxShadow="2px 2px 8px 0px #0000001A" alignItems="end">
            <CloseIcon data-testid="close-add-watchlist-modal" _hover={{ cursor: `pointer` }} marginRight="0.5rem" onClick={handleClose} />
            <Flex width="100%" padding="4px">
              <InputGroup>
                <InputLeftElement pointerEvents="none" color="gray.300" fontSize="1.2em">
                  <SearchIcon />
                </InputLeftElement>
                <Input data-testid="search-add-watchlist-modal" value={searchWatchlist} onChange={handleInputChange} variant="underlined" placeholder="Search" />
              </InputGroup>
            </Flex>
          </Flex>
          <ModalBody paddingX={0} maxH="300px" overflowY="scroll">
            <VStack divider={<Divider />} spacing="8px" alignItems="start">
              <WatchlistCheckbox watchlists={visibleWatchlists} warehouseProjectId={warehouseProjectId} />
            </VStack>
          </ModalBody>
          <ModalFooter boxShadow="2px 2px 8px 0px #0000001A" gap="48px" justifyContent="center">
            <Button fontWeight="500" rightIcon={<PlusIcon width="16px" height="16px" />} data-testid="search-add-create-modal" variant={`hoverOnly`} onClick={handleNewWatchlist}>
              {t(`newWatchlist`)}
            </Button>
          </ModalFooter>
        </ModalContent>
      </Modal>
      <EditWatchlistPopup isOpen={showCreatePopup} isNewWatchlist={true} title={t(`newWatchlist`)} onCancel={handleCreateCancel} onConfirm={handleCreateConfirm} />
    </>
  )
}
