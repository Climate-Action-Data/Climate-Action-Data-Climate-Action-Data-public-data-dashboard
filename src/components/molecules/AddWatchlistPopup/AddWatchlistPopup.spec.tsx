import { TestOvermindWrapper } from '@/test/TestOvermindWrapper'
import { AddWatchlistPopup } from './AddWatchlistPopup'
import { render, screen, fireEvent, waitFor } from '@testing-library/react'

const PROJECT_ID = `eb21d827-5a9c-43a5-97ff-a9e37e01a72a`
const MODAL_OPEN = true
const MODAL_CLOSED = false
const DELAY_FOR_RENDER = 1000

it(`renders correctly with modal open`, () => {
  const { container } = render(
    <TestOvermindWrapper>
      <AddWatchlistPopup isOpen={MODAL_OPEN} warehouseProjectId={PROJECT_ID} />
    </TestOvermindWrapper>,
  )
  expect(container).toMatchSnapshot()
})

it(`renders correctly with modal close`, () => {
  const { container } = render(
    <TestOvermindWrapper>
      <AddWatchlistPopup isOpen={MODAL_CLOSED} warehouseProjectId={PROJECT_ID} />
    </TestOvermindWrapper>,
  )
  expect(container).toMatchSnapshot()
})

it(`renders correctly and modal closes`, async () => {
  const mockModalClose = jest.fn()
  const { container } = render(
    <TestOvermindWrapper>
      <AddWatchlistPopup onModalClose={mockModalClose} isOpen={MODAL_OPEN} warehouseProjectId={PROJECT_ID} />
    </TestOvermindWrapper>,
  )
  await new Promise((r) => setTimeout(r, DELAY_FOR_RENDER))
  const closeButton = screen.getByTestId(`close-add-watchlist-modal`)
  fireEvent.click(closeButton)
  expect(mockModalClose).toHaveBeenCalled()
  expect(container).toMatchSnapshot()
})

it(`renders correctly and search modal opens`, async () => {
  const { container } = render(
    <TestOvermindWrapper>
      <AddWatchlistPopup isOpen={MODAL_OPEN} warehouseProjectId={PROJECT_ID} />
    </TestOvermindWrapper>,
  )
  await new Promise((r) => setTimeout(r, DELAY_FOR_RENDER))
  const closeButton = screen.getByTestId(`search-add-create-modal`)
  fireEvent.click(closeButton)
  expect(container).toMatchSnapshot()
})

it(`renders correctly and search modal opens and closes`, async () => {
  const { container } = render(
    <TestOvermindWrapper>
      <AddWatchlistPopup isOpen={MODAL_OPEN} warehouseProjectId={PROJECT_ID} />
    </TestOvermindWrapper>,
  )
  await new Promise((r) => setTimeout(r, DELAY_FOR_RENDER))
  const closeButton = screen.getByTestId(`search-add-create-modal`)
  fireEvent.click(closeButton)
  await new Promise((r) => setTimeout(r, DELAY_FOR_RENDER))
  fireEvent.click(screen.getByTestId(`cancel-modal`))
  expect(container).toMatchSnapshot()
})

it(`renders correctly and search modal opens and closes`, async () => {
  const { container } = render(
    <TestOvermindWrapper>
      <AddWatchlistPopup isOpen={MODAL_OPEN} warehouseProjectId={PROJECT_ID} />
    </TestOvermindWrapper>,
  )
  await new Promise((r) => setTimeout(r, DELAY_FOR_RENDER))
  const closeButton = screen.getByTestId(`search-add-create-modal`)
  fireEvent.click(closeButton)
  await waitFor(() => screen.getByTestId(`confirm-modal`))
  fireEvent.click(screen.getByTestId(`confirm-modal`))
  expect(container).toMatchSnapshot()
})

it(`renders correctly and search modal opens and closes`, async () => {
  const { container } = render(
    <TestOvermindWrapper>
      <AddWatchlistPopup isOpen={MODAL_OPEN} warehouseProjectId={PROJECT_ID} />
    </TestOvermindWrapper>,
  )

  await waitFor(() => screen.getByTestId(`search-add-watchlist-modal`))
  fireEvent.input(screen.getByTestId(`search-add-watchlist-modal`), { target: { value: `test` } })
  expect(container).toMatchSnapshot()
})

it(`renders correctly and search modal opens and closes`, async () => {
  const { container } = render(
    <TestOvermindWrapper>
      <AddWatchlistPopup isOpen={MODAL_OPEN} warehouseProjectId={PROJECT_ID} />
    </TestOvermindWrapper>,
  )

  await waitFor(() => screen.getByTestId(`search-add-watchlist-modal`))
  fireEvent.input(screen.getByTestId(`search-add-watchlist-modal`), { target: { value: `test` } })
  fireEvent.input(screen.getByTestId(`search-add-watchlist-modal`), { target: { value: `` } })
  expect(container).toMatchSnapshot()
})
