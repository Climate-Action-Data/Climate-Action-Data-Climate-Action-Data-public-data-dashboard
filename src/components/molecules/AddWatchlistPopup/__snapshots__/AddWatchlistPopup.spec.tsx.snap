// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly and modal closes 1`] = `<div />`;

exports[`renders correctly and search modal opens 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;

exports[`renders correctly and search modal opens and closes 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;

exports[`renders correctly and search modal opens and closes 2`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;

exports[`renders correctly and search modal opens and closes 3`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;

exports[`renders correctly and search modal opens and closes 4`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;

exports[`renders correctly with modal close 1`] = `<div />`;

exports[`renders correctly with modal open 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;
